File: python/198. House Robber.py
#coding=utf-8
__author__ = 'xuxuan'
class Solution(object):
    _map={}
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums)==0:
            return 0
        self._map={0:0,1:nums[-1]}
        return self.f(nums)
    def f(self,nums):
        n=len(nums)
        if self._map.get(n) is None:
            self._map[n]=max(nums[0]+self.f(nums[2:]),nums[1]+self.f(nums[3:]))
        return self._map[n]

File: java/198. House Robber.java
public class Solution {
    public int rob(int[] nums) {
        int n=nums.length;
        if (n==0) {
            return 0;
        }
        int[] ans=new int[n+1];
        ans[0]=0;
        ans[1]=nums[0];
        for (int i = 2; i < n+1; i++) {
            ans[i]=Math.max(nums[i-1]+ans[i-2], ans[i-1]);
        }
        return ans[n];
    }
}